File: Csharp/Lab8/Lab8/Lab8/cng/Model/Collections/deprecated/StackListImpl.cs
using System;
using System.Collections;
using System.Text;

namespace Lab8.cng.Model.Collections.deprecated
{
    public class StackListImpl <TParam> : ICustomCollections<TParam>
    {
        private ArrayList _stackArray = new ArrayList();

        public StackListImpl(){}


        public bool Add(TParam param)
        {
            _stackArray.Add(param);
            return true;
        }

        public TParam Get(int index)
        {
            return (TParam) _stackArray[index];
        }

        public int Size() {
            return _stackArray.Count;
        }

        
        public TParam Pop() {
            object returner = _stackArray[Size() - 1];
            _stackArray.Remove(Size() - 1);
            return (TParam) returner;
        }

        
        public bool IsEmpty() {
            return Size() == 0;
        }

        
        public Object[] CollectionToArray() {
            return _stackArray.ToArray();
        }

        public void SetElement(int index, TParam param)
        {
            _stackArray.Add(param);
        }


        public void SetElement(TParam param) {
            _stackArray.Add(param);
        }


        
        public override bool Equals(Object o) {
            return o != null && this.GetHashCode() == o.GetHashCode() && this.GetType() == o.GetType();
        }

        
        public override int GetHashCode() {
            const int constvar = 31;
            int result = 1;
            result = constvar * result + Size();
            foreach (TParam aStackArray in _stackArray) {
                result = constvar * result + aStackArray.GetHashCode();
            }
            return result;
        }

        
        public override String ToString() {
            StringBuilder sb = new StringBuilder().Append("Stack content:\n");
            foreach (TParam aStackArray in _stackArray) {
                if (aStackArray != null) {
                    sb.Append(aStackArray).Append("\n");
                } else {
                    sb.Append("Object is empty\n");
                }
            }
            return sb.ToString();
        }
    }
}

File: Kotlin/Lab8/src/by/bntu/fitr/povt/cng/model/collections/deprecated/StackListImpl.java
package by.bntu.fitr.povt.cng.model.collections.deprecated;

import java.util.*;

public class StackListImpl<T> implements Stack<T>, CustomCollections<T>, Iterable<T> {

    private final int INIT_SIZE = 16;
    private final int CUT_RATE = 4;
    private Object[] array = new Object[INIT_SIZE];
    private int pointer = 0;


    public StackListImpl() {

    }

    @SafeVarargs
    public StackListImpl(T... params) {
        for (T param : params) {
            add(param);
        }
    }

    public void push(T t) {
        add(t);
    }

    @Override
    public boolean add(T item) {
        if (pointer == array.length - 1){
            resize(array.length * 2);
        }
        array[pointer++] = item;
        return true;
    }

    public T pop() {
        T returner = (T) array[size() - 1];
        removeByIndex(size() - 1);
        return returner;
    }


    @Override
    public T get(int index) {
        return (T) array[index];
    }

    @Override
    public Object[] CollectionToArray() {
        Object[] objects = new Object[size()];
        System.arraycopy(array, 0, objects, 0, size());
        return objects;
    }

    @Deprecated
    @Override
    public void setElement(int index, T t) {

    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            int index = 0;

            @Override
            public boolean hasNext() {
                return index < pointer;
            }


            @Override
            public T next() {
                return (T) array[index++];
            }
        };
    }


    public Object[] toArray() {
        Object[] objects = new Object[size()];
        System.arraycopy(array, 0, objects, 0, size());
        return objects;
    }

    @Override
    public synchronized boolean isEmpty() {
        return pointer == 0;
    }


    private void removeByIndex(int index) {
        System.arraycopy(array, index + 1, array, index, pointer - index);
        array[pointer] = null;
        pointer--;
        if (array.length > INIT_SIZE && pointer < array.length / CUT_RATE) {
            resize(array.length / 2);
        }


    }

    @Override
    public int size() {
        return pointer;
    }

    private void resize(int newLength) {
        Object[] newArray = new Object[newLength];
        System.arraycopy(array, 0, newArray, 0, pointer);
        array = newArray;
    }


    @Override
    public boolean equals(Object o) {
        return o != null && this.hashCode() == o.hashCode() && this.getClass() == o.getClass();
    }

    @Override
    public int hashCode() {
        final int CONSTVAR = 31;
        int result = 1;
        result = CONSTVAR * result + size();
        for (int i = 0; i < size(); i++) {
            result = CONSTVAR * result + array[i].hashCode();
        }
        return result;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder().append("Stack content:\n");
        for (Object a : array) {
            if (a != null) {
                sb.append(a.toString()).append("\n");
            } else {
                sb.append("Object is empty\n");
            }
        }
        return sb.toString();
    }

}
